### PR TITLE
[DmlPrototype] Introduce new optimizer MatMul + BatchNormalization

### PR DESCRIPTION
**Pull request by sumitsays**
### Description
Introduce new ORT L1 optimizer under RewriteRule category to fuse MatMul + BatchNormalization node. This optimizer look for a specific pattern observed in one of the impacting customer models and fuse the Matmul and Batchnormalization node into a Gemm node. For details on the pattern matching and fusion please refer to the comment section of `matmul_bn_fusion.cc`.

To visualize this optimizer will perform following transformation.

image

Note: This optimizer may evolve in the future to be more generic in terms of the pattern matching.

### Motivation and Context
- Why is this change required? What problem does it solve?
One of the user of ORT+DML ep needs this to better target the model to DML. But this transformation applies more broadly, so added L1 optimizer.

### CODE REVIEW

**Comment by sumitsays** [comment]
I think ORT main is broken because there are build failures in `onnxruntime/core/providers/rocm/nn/batch_norm.h` for other  PRs as well. Therefore creating a copy of the PR to DmlPrototype branch, as we want to unblock Intel.
Also I am not sure whether any ORT folks will review the non-main PR, so not adding them for this PR. Whenever I will cherry-pick this PR to ORT main, I will add ORT folks to get it reviewed.
